compiler/rustc_span/src/source_map.rs

```diff
@@ -115,13 +116,18 @@ impl FileLoader for RealFileLoader {
     }
 
     fn read_file(&self, path: &Path) -> io::Result<String> {
-        if path.metadata().is_ok_and(|metadata| metadata.len() > SourceFile::MAX_FILE_SIZE.into()) {
+        let mut file = File::open(path)?;
+        let size = file.metadata().map(|metadata| metadata.len()).ok().unwrap_or(0);
+
+        if size > SourceFile::MAX_FILE_SIZE.into() {
             return Err(io::Error::other(format!(
                 "text files larger than {} bytes are unsupported",
                 SourceFile::MAX_FILE_SIZE
             )));
         }
-        fs::read_to_string(path)
+        let mut contents = String::with_capacity(size as usize);
```

Comment:
This isn't actually needed, file.read_to_string also calls try_reserve which has better error-handling in OOM situations.

Reply:
Ah, I thought it doesn't matter where we reserve, but I didn't realize that this would panic, while the stdlib impl returns an error. Good catch!